### DISCUSSION TITLE
Ideas for dynamic management of mocks?

### DISCUSSION THREAD

**Discussion by joeattardi** [mswjs/msw#1533] [Q&A] [answered]
I am looking into wider adoption of MSW in my project but am not sure the best way to proceed.

Typically, in the browser, my team only really uses mocks while developing a specific feature while we wait for the backend team to implement the API changes we need. After the backend support is done for a feature, we typically don't use that mock anymore. So I am trying to come up with a way to have short lived mock handlers. Our repository is already pretty large and we want to avoid cluttering it with many mock files, most of which won't be used after a certain feature is done.

Are there any good strategies out there for this? The way we currently do mocking is a bit limited. We have static JSON files that are managed by a local development proxy. The upside to this is that it's easy to dynamically manage these. The dev proxy server watches a `mocks` directory and when it sees JSON files in there it dynamically loads them and defines the mock. We have a rule in `.gitignore` so that these mocks aren't committed to source control.

Then typically when we submit a pull request, the mock JSON is attached so that people can use it for testing.

Since in the browser we can't watch a directory for files, was wondering if there were any good patterns for something like this with MSW where we can dynamically register a given mock?

**Comment by kettanaito** [comment, marked as answer]
Hey, @joeattardi. Thanks for raising this.

I think keeping a local proxy server is a bit redundant if you're planning on using MSW. The library can give you a much better control over the network without having to manage servers. You can certainly generate request handlers for MSW from any source, including static JSON files, but I'd first recommend trying out plain handlers as they give you freedom to capture dynamic scenarios, like conditional responses, whereas JSON is always static.

## Mindset

Here's a mindset that can help you find the right setup with MSW: I encourage people to mock _endpoints_ and then encapsulate any branching logic of the endpoint within a single handler. Something like this:

```js
rest.get('/user', (req, res, ctx) => {
  const userId = req.url.searchParams.get('id')
  const mockUser = users.get(userId)

  if (!mockUser) {
    return res(ctx.status(404))
  }

  return res(ctx.json(mockUser))
})
```

If this gets verbose over time, I then recommend people grouping handlers by endpoint and defining conditional behaviors using handler overrides:

```js
// mocks/user.js
export const handlers = [
  rest.get('/user', (req, res, ctx) => {
    if (!users.has(req.url.searchParams.get('id')) {
      return res(ctx.status(404))
    }

    // Otherwise do nothing, fall deeper into other handlers.
    return req.passthrough()
  }),
  rest.get('/user', (req, res, ctx) => {
    return res(ctx.json(users.get(req.url.searchParams.get('id')))
  })
]
```

Keeping in mind that you can combine handlers per endpoints wherever you need:

```js
// mocks.js
import * as userMocks from './user'

export const handlers = [...userMocks]
```

## General statements

Overall, I'd recommend committing the mocks to Git. This way everybody has a clear picture of what's mocked and you have a clear history to track changes to those mocks over time. Everybody wins.

I'd also recommend keeping the mocking setup as simple as possible even if that includes repeating yourself. Mocks often have 0 maintenance cost, so designing a complicated setup around them only hurts discoverability and debugging in the long run. That being said, if you and your team are completely on board with the setup, you can introduce pretty much anything that'd suit your needs (e.g. keeping the list of endpoints, generating handlers out of them, then writing an automation job that would ping those endpoints and notify you when respective mocks can be removed). This is really up to you, MSW doesn't enforce nor forbids any particular setup.

**Comment by joeattardi** [comment]
Thanks for your reply!

I agree about the proxy being redundant with MSW; my hope is we can remove at least the mocking piece of the proxy (it's a server we run locally that proxies requests to the real API, though I had an earlier question about if MSW can replace that functionality too and it seems that it may be an option as well!)

Keeping the mocks in git isn't a bad idea; we have definitely had times where a mock was accidentally wiped out and since it was ignored, could not be recovered.

I was thinking about ways to have the mocks remain in the git repository but control which mocks are active. Here's a thought - what about using local storage? We can expose some functions in development mode that can register which mocks to use, and use local storage to persist that across page reloads.

That way the mocks can remain, but won't be activated unless needed. Manually activating them in the console takes care of the problem of discovering them dynamically, but can mocks be registered after the worker starts? Or would we have to mark which mocks to use, and then refresh the page?

My thinking was something like, I load up my local UI and decide I want to use a mock. I open my console and run something like (random syntax idea)

```
useMock('/path/to/api/mock');
```

It would register in local storage that want to use that mock. Now, could we dynamically import this mock and add it to the active handlers? Or would a page refresh be needed?

Thanks again!

**Reply by kettanaito** [reply]
Totally! That shouldn't be that complex to do using `worker.use()` on runtime. 

> Now, could we dynamically import this mock and add it to the active handlers? Or would a page refresh be needed?

I believe you can't issue any kind of imports on runtime. The file system doesn't exist in the browser so it wouldn't resolve paths to mock files anyway. 

What you can do is keep all the mocks in memory, and then apply them via `worker.use()` called in your `useMock()` function. No page reload would be necessary, MSW handlers take effect immediately because they are, too, stored in memory. 

You can utilize `localStorage` to persist your previous choices. As I said, your imagination is the only limit. Just don't make this setup too smart.

**Reply by kettanaito** [reply]
I know there are a few projects in the community to visualize MSW and enable behaviors that are very similar to what you're trying to build. I could recall only [Switchboard](https://github.com/coryhouse/switchboard) at the moment but you can look around.

**Reply by joeattardi** [reply]
> I believe you can't issue any kind of imports on runtime. The file system doesn't exist in the browser so it wouldn't resolve paths to mock files anyway.

Ah, I think you're right. I was thinking a Webpack dynamic import, but I forgot that the path given to an `import(...)` call has to be a static path and can't be a variable (that has bitten me before).

Maybe loading them all and just not registering them initially is the best way to go!

**Reply by kettanaito** [reply]
Yeah, just keep them in memory and only use `useMock` to call `worker.use()` with the correct handlers by endpoint path, name, internal ID—anything that suits you.

**Reply by joeattardi** [reply]
This is fantastic. Thanks yet again for always providing great help here!

**Reply by kettanaito** [reply]
Hey, if you like MSW please consider [sponsoring the project](https://github.com/sponsors/mswjs). Absolutely every contribution counts as I'm maintaining this colossus for free in my spare time. Ideally, this is a great topic to raise with your employer. 

In any case, thank you!
